l1-contracts/contracts/governance/PermanentRestriction.sol

```diff
@@ -0,0 +1,189 @@
+            revert CallNotAllowed(_call.data);
+        }
+    }
+
+    /// @notice Validates the correctness of the new admin.
+    /// @param _call The call data.
+    /// @dev Ensures that the admin has a whitelisted implementation and does not remove this restriction.
+    function _validateNewAdmin(Call calldata _call) internal view {
+        address newChainAdmin = abi.decode(_call.data[4:], (address));
+
+        bytes32 implementationCodeHash;
+        assembly {
+            implementationCodeHash := extcodehash(newChainAdmin)
+        }
+
+        if(!allowedAdminImplementations[implementationCodeHash]) {
+            revert UnallowedImplementatioN(implementationCodeHash);
+        }
+
+        // Since the implementation is known to be correct (from the checks above), we
+        // can safely trust the returned value from the call below
+        if (!IChainAdmin(newChainAdmin).isRestrictionActive(address(this))) {
+            revert RemovingPermanentRestriction();
+        }
+    }
+
+    /// @notice Validates the removal of the restriction.
+    /// @param _call The call data.
+    /// @dev Ensures that this restriction is not removed.
+    function _validateRemoveRestriction(Call calldata _call) internal view {
+        if (_call.target != msg.sender) {
+            return;
+        }
+
+        if (bytes4(_call.data[:4]) != IChainAdmin.removeRestriction.selector) {
+            return;
+        }
+
+        address removedRestriction = abi.decode(_call.data[4:], (address));
+
+        if (removedRestriction == address(this)) {
+            revert RemovingPermanentRestriction();
+        }
+    }
+
+    /// @notice Checks if the `msg.sender` is an admin of a certain ZkSyncHyperchain.
+    /// @param _chain The address of the chain.
+    function _isAdminOfAChain(address _chain) internal view returns (bool) {
+        (bool success, ) = address(this).staticcall(abi.encodeCall(this.tryCompareAdminOfAChain, (_chain, msg.sender)));
+        return success;
+    }
+
+    /// @notice Tries to compare the admin of a chain with the potential admin.
+    /// @param _chain The address of the chain.
+    /// @param _potentialAdmin The address of the potential admin.
+    /// @dev This function reverts if the `_chain` is not a ZkSyncHyperchain or the `_potentialAdmin` is not the
+    /// admin of the chain.
+    function tryCompareAdminOfAChain(address _chain, address _potentialAdmin) external view {
+        if (_chain == address(0)) {
+            revert ChainZeroAddress();
+        }
+
+        // Unfortunately there is no easy way to double check that indeed the `_chain` is a ZkSyncHyperchain.
+        // So we do the following:
+        // - Query it for `chainId`. If it reverts, it is not a ZkSyncHyperchain.
+        // - Query the Bridgehub for the Hyperchain with the given `chainId`.
+        // - We compare the corresponding addresses
+        uint256 chainId = IZkSyncHyperchain(_chain).getChainId();
```

Comment:
I can't find this method in `IZkSyncHyperchain`, what I am missing?

Reply:
I think it was removed when I rebased to dev, I'll add it ehre